gax/src/main/java/com/google/api/gax/batching/BatcherImpl.java

```diff
@@ -127,8 +171,29 @@ public BatcherImpl(
   @Override
   public ApiFuture<ElementResultT> add(ElementT element) {
     Preconditions.checkState(!isClosed, "Cannot add elements on a closed batcher");
-    SettableApiFuture<ElementResultT> result = SettableApiFuture.create();
+    // This is not the optimal way of throttling. It does not send out partial batches, which
+    // means that the Batcher might not use up all the resources allowed by FlowController.
+    // The more efficient implementation should look like:
+    // if (!flowController.tryReserve(1, bytes)) {
+    //   sendOutstanding();
+    //   reserve(1, bytes);
+    // }
+    // where tryReserve() will return false if there isn't enough resources, or reserve and return
+    // true.
+    // However, with the current FlowController implementation, adding a tryReserve() could be
+    // confusing. FlowController will end up having 3 different reserve behaviors: blocking,
+    // non blocking and try reserve. And we'll also need to add a tryAcquire() to the Semaphore64
+    // class, which made it seem unnecessary to have blocking and non-blocking semaphore
+    // implementations. Some refactoring may be needed for the optimized implementation. So we'll
+    // defer it till we decide on if refactoring FlowController is necessary.
+    try {
+      flowController.reserve(1, batchingDescriptor.countBytes(element));
```

Comment:
Just wondering, how heavy can `batchingDescriptoro.countBytes()` get? I see in bigtable implementation it is just `ByteString.size()` which i assume is O(1), but we need to be careful here, to avoid some potential unnecessary serializations to bytes of complex objects just to calculate the size of it (and then throw serialization results away, just to be repeated again on actual send).

Reply:
I don’t think the results are thrown away...I think the protobuf message will serialize and cache the serialized representation so it has less work to do in the future